File: backend/app/grid.py
from math import cos, radians

CELL_KM = 0.1
KM_PER_DEG_LAT = 111.0
COS_LAT = cos(radians(13.0))
LAT_STEP = CELL_KM / KM_PER_DEG_LAT
LON_STEP = CELL_KM / (KM_PER_DEG_LAT * COS_LAT)


def point_to_cell_id(lat: float, lon: float) -> str:
    i = round(lat / LAT_STEP)
    j = round(lon / LON_STEP)
    return f"{i}_{j}"


def cell_id_to_center(cell_id: str) -> tuple[float, float]:
    i_str, j_str = cell_id.split("_")
    return int(i_str) * LAT_STEP, int(j_str) * LON_STEP

File: frontend/src/lib/grid.ts
const CELL_KM = 0.3;
const KM_PER_DEG_LAT = 111.0;
const COS_LAT = Math.cos((13.0 * Math.PI) / 180);
const LAT_STEP = CELL_KM / KM_PER_DEG_LAT;
const LON_STEP = CELL_KM / (KM_PER_DEG_LAT * COS_LAT);

export const cellCenter = (cellId: string): [number, number] => {
  const [i, j] = cellId.split('_').map(Number);
  return [i * LAT_STEP, j * LON_STEP];
};

export type CellBounds = [[number, number], [number, number], [number, number], [number, number]];

export const cellBounds = (cellId: string): CellBounds => {
  const [lat, lon] = cellCenter(cellId);
  const dLat = LAT_STEP / 2;
  const dLon = LON_STEP / 2;
  return [
    [lon - dLon, lat - dLat],
    [lon + dLon, lat - dLat],
    [lon + dLon, lat + dLat],
    [lon - dLon, lat + dLat],
  ];
};

export const cellToGeoJSON = (cellId: string) => {
  const bounds = cellBounds(cellId);
  return {
    type: 'Feature' as const,
    properties: { cellId },
    geometry: {
      type: 'Polygon' as const,
      coordinates: [[...bounds, bounds[0]]],
    },
  };
};
